### ISSUE TITLE
Swift UI Integration

### ISSUE THREAD

**Issue by djshumpert** [stripe/stripe-ios#1204]
Hi, I recently started a Swift UI project and was trying to integrate creating customer tokens within the app. Has anyone tried integrating Stripe with Swift UI? I know it is still in beta 🙈

**Comment by yuki-stripe** [comment]
Hey @djshumpert, we have not - interested to hear what you find!

**Comment by patchthecode** [comment]
@djshumpert did you make any progress? I am trying the same thing

**Comment by amuresia** [comment]
Hi all, I've given it a go in the last couple of days but now with the SCA in place, I get stuck when trying to [support 3D Secure Authentication on iOS](https://stripe.com/docs/mobile/ios/authentication#implement-authentication-context)

I can't seem to be able to implement the STPAuthenticationContext protocol as my application doesn't use any UIViewControllers. Is anybody aware of any workarounds?

@yuki-stripe could we please re-open this issue as no solution was provided.

**Comment by Gujci** [comment]
@amuresia  I have just ran into the same issue. It's sad, that `STPAuthenticationContext` needs to be an `NSObject`. Sadly, I have to go back to the old `ViewController` mess, and maybe wrap it into a `UIViewControllerRepresentable`. Right now I see no other option to provide the context.

**Comment by davidme-stripe** [comment]
Thanks for the feedback! We'll see if we can make this less messy.

**Comment by davidme-stripe** [comment]
I need to wrap my head around this a bit more, but it looks like SwiftUI apps still have a root UIViewController (generally the UIHostingController), so you should still be able to present legacy view controllers without too many modifications to your app. Could you do something like this?

```
struct MyView: View {
    private class PaymentContextDelegate: NSObject, STPPaymentContextDelegate {
        /* Implement STPPaymentContextDelegate methods */
    }
   
    private var paymentContextDelegate = PaymentContextDelegate()

    var body: some View {
        VStack(alignment: .leading) {
            Text("SwiftUI Test")
            Button(action: {
                let customerContext = STPCustomerContext(keyProvider: MyAPIClient.sharedClient)
                let context = STPPaymentContext(customerContext: customerContext)
                let rootVC = UIApplication.shared.keyWindow?.rootViewController // You probably shouldn't do this to get the rootViewController if you support multiple UIScenes: You'll want some way to know which scene the MyView is hosted in.
                context.hostViewController = rootVC
                context.delegate = self.paymentContextDelegate
                context.paymentAmount = 1000
                context.requestPayment()
            }) {
                Text("Pay")
            }
        }
    }
}
```

Let me know if this unblocks you for now. It obviously isn't ideal, and we'll continue to investigate ways to provide a more native integration for SwiftUI.

**Comment by amuresia** [comment]
Hey @davidme-stripe! Thanks for the solution - it's definitely a step forward at least for now.
Could you please also add a snippet of the way you implement and share your MyAPIClient.sharedClient? What type does it have because I am trying to follow step 2 from this tutorial https://stripe.com/docs/mobile/ios/standard#ephemeral-key but I'm a bit confused because it suggests I should cast my JSON response to [String: AnyObject] however it always fails to cast the response.

I tried the following approaches:
```
# Compiler error: Cast from 'String' to unrelated type '[String : AnyObject]' always fails
                let result = task.result!
                let responseString = String(data: result.responseData!, encoding: .utf8)
                completion(responseString! as? [String: AnyObject], nil)
```

```
# Compiler error:Cast from 'Data?' to unrelated type '[String : AnyObject]' always fails
                let result = task.result!
                completion(task.result!.responseData as? [String: AnyObject], nil)
```

And in both cases the View complains:
```
Argument type '[AnyHashable : Any]?' does not conform to expected type 'STPCustomerEphemeralKeyProvider'```
unless I cast it again ...
```                    let customerContext = STPCustomerContext(keyProvider: key as! STPCustomerEphemeralKeyProvider )
```

**Comment by davidme-stripe** [comment]
Try looking at [MyAPIClient.swift](https://github.com/stripe/stripe-ios/blob/c26fffa372d882da12bd6a677978b1a2892be956/Example/Standard%20Integration/MyAPIClient.swift) in this repository, which was built to communicate with our [example-ios-backend](https://github.com/stripe/example-ios-backend/tree/v17.0.0). The APIClient should conform to `STPCustomerEphemeralKeyProvider`, which involves implementing `createCustomerKey(apiVersion:, completion:)`. If you then pass in your API Client into `STPCustomerContext` as the `keyProvider`, we'll call `createCustomerKey()` on it when we need to create the key.

**Comment by amuresia** [comment]
Thank you very much kind sir! I've now got a working PoC which I'll continue to build on top of. I'll make sure to update this issue if some 🐛 crops up!

**Comment by Gujci** [comment]
@dackerman-stripe excuse me for falsely using the word "mess", Monday was quite exhausting for me. 

The problem in may case was the following:

- made a custom swiftUI view, using your `STPPaymentCardTextField` wrapped into StripePaymentCardTextField, which works really nice 👍 
- this custom view had a view model, with a `@Published` `STPPaymentMethodCardParams`, this was bonded to the views' input.
- On my custom save button, I started to make all the calls needed to register a card
- Then came the last step, where I faced, that I would need a ViewController instance in my view model somehow. (ViewModels must not store view references, but you all know that, I'm just clarifying)

One solution I can imagine is having a `UIHostingController` subclass / make `UIHostingController ` conform to `STPAuthenticationContext`, which may wrap my SwiftUI view. Then, this `UIHostingController` can be further wrapped into a `UIViewControllerRepresentable` to make it usable from SwiftUI. This wrapper would take the view model in as a parameter, add the `UIHostingController` reference to the view model (which is the main low in this idea) and pass the viewModel further to my custom SwiftUI view.

Your solution is somewhat similar at the point, where we need some reference to a UIViewController. In your case it's through a singleton.

_To achieve a cleaner solution, right now I can only imagine having to modify the requirements of `STPAuthenticationContext`, so a non UIKit type can adopt it. It might flip all the internals of the framework, so, I don't think this solution could work. Of course it's your framework, so I'm in no position to make assumptions._

**Comment by Gujci** [comment]
Ended up subclassing `UIHostingController`, since generic extensions are not quite good with objc protocols.

```swift
class CardRegisterHostingController: UIHostingController<MySwiftUIView>, STPAuthenticationContext { ... }
```

**Comment by amuresia** [comment]
HI @davidme-stripe,

Not sure if it's related to the integration with SwiftUI but when I call
```paymentContext.presentShippingViewController()```
as [suggested in step 5 of the docs](https://stripe.com/docs/mobile/ios/standard#set-up-customer-context), the view that is pushed on top doesn't have any input fields for the shipping address and looks like this
<img width="409" alt="image" src="https://user-images.githubusercontent.com/16068752/67164302-1e7eef00-f371-11e9-9808-f30f91bd4301.png">
If I press 'Next' then things look better and I get the logic included in 
```
paymentContext(_ paymentContext: STPPaymentContext, didUpdateShippingAddress address: STPAddress, completion: @escaping STPShippingMethodsCompletionBlock) {
```
<img width="409" alt="image" src="https://user-images.githubusercontent.com/16068752/67164332-7584c400-f371-11e9-9dc5-3b17faced537.png">

Would you be so kind to tell me if I am missing some configuration or if there's a bug/ limitation that I stumbled across? 

Thanks,
Andrei

**Comment by lukeharkin1993** [comment]
> HI @davidme-stripe,
> 
> Not sure if it's related to the integration with SwiftUI but when I call
> `paymentContext.presentShippingViewController()`
> as [suggested in step 5 of the docs](https://stripe.com/docs/mobile/ios/standard#set-up-customer-context), the view that is pushed on top doesn't have any input fields for the shipping address and looks like this
> <img alt="image" width="409" src="https://user-images.githubusercontent.com/16068752/67164302-1e7eef00-f371-11e9-9808-f30f91bd4301.png">
> If I press 'Next' then things look better and I get the logic included in
> 
> ```
> paymentContext(_ paymentContext: STPPaymentContext, didUpdateShippingAddress address: STPAddress, completion: @escaping STPShippingMethodsCompletionBlock) {
> ```
> 
> <img alt="image" width="409" src="https://user-images.githubusercontent.com/16068752/67164332-7584c400-f371-11e9-9dc5-3b17faced537.png">
> 
> Would you be so kind to tell me if I am missing some configuration or if there's a bug/ limitation that I stumbled across?
> 
> Thanks,
> Andrei

Hey Andrei,

Did you ever solve this? I am having the same issue too.

Thanks,
Luke

**Comment by amuresia** [comment]
Hey @lukeharkin1993, I am afraid not ... I had to implement my own shipping logic/ views :(

**Comment by lukeharkin1993** [comment]
> Hey @lukeharkin1993, I am afraid not ... I had to implement my own shipping logic/ views :(

Ah no worries @amuresia, there is a - containsRequiredShippingAddressFields, but it doesn't seem to make a difference :/ thanks anyway. I think I will follow the same route you did.

**Comment by davidme-stripe** [comment]
Hi @amuresia and @lukeharkin1993, very sorry that I missed this!

It looks like you may not have set `requiredShippingAddressFields` in your STPPaymentConfiguration, so the STPShippingViewController has nothing to collect. Try something like this: `paymentContext.configuration.requiredShippingAddressFields = [.name, .postalAddress, .phoneNumber, .emailAddress]`. We'll look into making this more obvious in a future update.

If you have any problems in the future, please open a new GitHub issue or reach out to https://support.stripe.com — it'll help us track your request and get you a quicker response!

**Comment by davidme-stripe** [comment]
Hi @amuresia and @lukeharkin1993, very sorry that I missed this!

It looks like you may not have set `requiredShippingAddressFields` in your STPPaymentConfiguration, so the STPShippingViewController has nothing to collect. Try something like this: `paymentContext.configuration.requiredShippingAddressFields = [.name, .postalAddress, .phoneNumber, .emailAddress]`. We'll look into making this more obvious in a future update.

If you have any problems in the future, please open a new GitHub issue or reach out to https://support.stripe.com — it'll help us track your request and get you a quicker response!

**Comment by lukeharkin1993** [comment]
Hi @davidme-stripe , I have implemented what you suggested. It works great thanks.
Unfortunately when I call `presentShippingViewController()` after adding a card by calling `presentPaymentOptionsViewController()`, `didCreatePaymentResult()` never gets called. If I bypass `presentShippingViewController()` and just call `requestPayment()` straight after adding a card then the payment goes through fine. Have you any idea what might be happening? I can't find anything in the log and paymentContext doesn't seem to be nil. Thanks.

**Comment by davidme-stripe** [comment]
Hey @lukeharkin1993, just to clarify: Are you calling `presentShippingViewController()`, then `requestPayment()` once the user has entered their shipping information, and not seeing `didCreatePaymentResult()` get called? It's expected that if you're not calling `requestPayment()`, it will never create the payment result.

If you're just trying to show the shipping view controller as part of your payment flow, it should work to just set `requiredShippingAddressFields` then call `requestPayment()` — STPPaymentContext will display the shipping view controller if it needs to.

**Comment by lukeharkin1993** [comment]
So initially what I was doing was calling  `presentPaymentOptionsViewController()` then setting `requiredShippingAddressFields` and then calling-> `presentShippingViewController()` then -> `requestPayment()`.

What I have gathered from your comment is I should be setting `requiredShippingAddressFields` then calling->`presentPaymentOptionsViewController()` then-> `requestPayment()` which then automatically presents the shipping view controller. Once I enter the details and hit next...it doesn't seem to reach `didCreatePaymentResult`

**Comment by davidme-stripe** [comment]
Ah, both of those should work correctly and reach `didCreatePaymentResult`. This sounds like a different bug, I'm going to move it out to a new issue so we can track it better.

**Comment by btian** [comment]
What's the right way to present STPPaymentOptionsViewController in SwiftUI?

I'm doing something like

```
let keyWindow = UIApplication.shared.connectedScenes
            .filter({$0.activationState == .foregroundActive})
            .map({$0 as? UIWindowScene})
            .compactMap({$0})
            .first?.windows
            .filter({$0.isKeyWindow}).first
        self.paymentContext.hostViewController = keyWindow?.rootViewController

Button(action: {
                self.paymentContext.presentPaymentOptionsViewController()
            }) {
                Text(self.paymentContextDelegate.paymentMethodButtonTitle)
```

But selecting a card or clicking the cancel button don't dismiss the view.

**Comment by amuresia** [comment]
Hey everyone! It's been a while since this issue was first brought up.
I was able to successfully integrate stripe last year, back when I was using the early beta version of SwiftUI. Let's just say that the app didn't make the cut.

I started working on a new app in my free time and I am using the latest tool available, I think Apple refers to it as SwiftUI version 2 ... the only relevant change for the stripe sdk is that the app no longer uses an AppDelegate and SceneDelegate.

I have been fighting with it for 2 days now and I am still blocked. I raised a ticket with the support team and I am waiting to hear back from them via email.

Essentially, I can allow the user to select a payment method by calling the method `presentPaymentOptionsViewController()`. So far so good. I then proceed with my workflow and call `requestPayment()` and this is when the app crashes complaining that the customer context is nil.

After debugging for hours in a row I noticed that when the screen from `presentPaymentOptionsViewController()` is dismissed, all of the screens are reinitialised 
I am using the same approach as @btian to get to the root view controller
```
let keyWindow = UIApplication.shared.connectedScenes
            .filter({$0.activationState == .foregroundActive})
            .map({$0 as? UIWindowScene})
            .compactMap({$0})
            .first?.windows
            .filter({$0.isKeyWindow}).first
        self.paymentContext.hostViewController = keyWindow?.rootViewController
```
I tried with `.filter({$0.isKeyWindow}).last` which doesn't crash the app but I am unable to type anything in the form.

What is interesting is that if I open the app again (re-build in XCode for example) and call `requestPayment()` **without previously calling** `presentPaymentOptionsViewController()` then I see a fulfilled payment intent in the dashboard.

Bottom line is that I can call `presentPaymentOptionsViewController` or `requestPayment` but not sequentially like I am supposed to (as indicated by the docs as well) because when one screen is dismissed, everything in the app reloads, causing the delegate and the customer context to be reinitialised.

I tried everything that I could think of to get over this and bypass the reinitialisation that occurs but I can't because there is  no AppDelegate or SceneDelegate in SwiftUI 2.

Any help would be appreciated. Thank you!

**Comment by mrkvans** [comment]
I'm surprised Stripe has ignored this issue for over a year now. The solution for iOS13 should be the same for iOS14 - so where is the iOS13 support for SwiftUI? Where are Stripe's plans/roadmap to support Apple's multi-platform (Apple platforms) App development framework? With all the time and resources that Stripe Puts into its documentation, the massive gap in support here is baffling to me.

**Comment by bat111** [comment]
> Hey everyone! It's been a while since this issue was first brought up.
> I was able to successfully integrate stripe last year, back when I was using the early beta version of SwiftUI. Let's just say that the app didn't make the cut.
> 
> I started working on a new app in my free time and I am using the latest tool available, I think Apple refers to it as SwiftUI version 2 ... the only relevant change for the stripe sdk is that the app no longer uses an AppDelegate and SceneDelegate.
> 
> I have been fighting with it for 2 days now and I am still blocked. I raised a ticket with the support team and I am waiting to hear back from them via email.
> 
> Essentially, I can allow the user to select a payment method by calling the method `presentPaymentOptionsViewController()`. So far so good. I then proceed with my workflow and call `requestPayment()` and this is when the app crashes complaining that the customer context is nil.
> 
> After debugging for hours in a row I noticed that when the screen from `presentPaymentOptionsViewController()` is dismissed, all of the screens are reinitialised
> I am using the same approach as @btian to get to the root view controller
> 
> ```
> let keyWindow = UIApplication.shared.connectedScenes
>             .filter({$0.activationState == .foregroundActive})
>             .map({$0 as? UIWindowScene})
>             .compactMap({$0})
>             .first?.windows
>             .filter({$0.isKeyWindow}).first
>         self.paymentContext.hostViewController = keyWindow?.rootViewController
> ```
> 
> I tried with `.filter({$0.isKeyWindow}).last` which doesn't crash the app but I am unable to type anything in the form.
> 
> What is interesting is that if I open the app again (re-build in XCode for example) and call `requestPayment()` **without previously calling** `presentPaymentOptionsViewController()` then I see a fulfilled payment intent in the dashboard.
> 
> Bottom line is that I can call `presentPaymentOptionsViewController` or `requestPayment` but not sequentially like I am supposed to (as indicated by the docs as well) because when one screen is dismissed, everything in the app reloads, causing the delegate and the customer context to be reinitialised.
> 
> I tried everything that I could think of to get over this and bypass the reinitialisation that occurs but I can't because there is no AppDelegate or SceneDelegate in SwiftUI 2.
> 
> Any help would be appreciated. Thank you!

hi amuresia 

i am working to build my app on swiftui 1.0.
i am glad you figured out to manage integrate Stripe into SwiftUi. 

> Hey everyone! It's been a while since this issue was first brought up.
> I was able to successfully integrate stripe last year, back when I was using the early beta version of SwiftUI. Let's just say that the app didn't make the cut.
> 
> I started working on a new app in my free time and I am using the latest tool available, I think Apple refers to it as SwiftUI version 2 ... the only relevant change for the stripe sdk is that the app no longer uses an AppDelegate and SceneDelegate.
> 
> I have been fighting with it for 2 days now and I am still blocked. I raised a ticket with the support team and I am waiting to hear back from them via email.
> 
> Essentially, I can allow the user to select a payment method by calling the method `presentPaymentOptionsViewController()`. So far so good. I then proceed with my workflow and call `requestPayment()` and this is when the app crashes complaining that the customer context is nil.
> 
> After debugging for hours in a row I noticed that when the screen from `presentPaymentOptionsViewController()` is dismissed, all of the screens are reinitialised
> I am using the same approach as @btian to get to the root view controller
> 
> ```
> let keyWindow = UIApplication.shared.connectedScenes
>             .filter({$0.activationState == .foregroundActive})
>             .map({$0 as? UIWindowScene})
>             .compactMap({$0})
>             .first?.windows
>             .filter({$0.isKeyWindow}).first
>         self.paymentContext.hostViewController = keyWindow?.rootViewController
> ```
> 
> I tried with `.filter({$0.isKeyWindow}).last` which doesn't crash the app but I am unable to type anything in the form.
> 
> What is interesting is that if I open the app again (re-build in XCode for example) and call `requestPayment()` **without previously calling** `presentPaymentOptionsViewController()` then I see a fulfilled payment intent in the dashboard.
> 
> Bottom line is that I can call `presentPaymentOptionsViewController` or `requestPayment` but not sequentially like I am supposed to (as indicated by the docs as well) because when one screen is dismissed, everything in the app reloads, causing the delegate and the customer context to be reinitialised.
> 
> I tried everything that I could think of to get over this and bypass the reinitialisation that occurs but I can't because there is no AppDelegate or SceneDelegate in SwiftUI 2.
> 
> Any help would be appreciated. Thank you!

Can you put up your ready version on github helping other developer to look at it.

Bat

**Comment by mrkvans** [comment]
@david-stripe Has your team looked at this recently? It would be nice to have native SwiftUI support, especially since all the Stipe UIViewControllers are called and managed by the Stripe Frameworks. 

At the very least, could you give us some samples of UIViewControllerRepresentable code that could be exposed in SwiftUI views? 

Or, extra credit if you provide sample code for 100% ApplePay UI using Stripe processing.

**Comment by daihovey** [comment]
I ended up making a fake SwiftUi StripeView and setting the frame to zero and then using the stripe calls from inside the FakeStripeViewController. Need to play around with it bit more, but seems to work; and 3DS modal works too.

```
import SwiftUI
import Stripe
import UIKit

struct StripeBuyView: View {
    @EnvironmentObject var viewModel: ViewModel
    @State var isPressed = false
    var body: some View {
        VStack {
            Button(action: {
                isPressed = !isPressed
            }, label: {
                Text("Buy")
            })
           
            FakeStripeView(isPressed: $isPressed)
                .frame(width: 0, height: 0)
        }
    }
}

struct FakeStripeView: UIViewControllerRepresentable {
    @EnvironmentObject var viewModel: ViewModel
    @Binding var isPressed: Bool
    
    public typealias UIViewControllerType = FakeStripeViewController
    
    func makeUIViewController(context: UIViewControllerRepresentableContext<FakeStripeView>) -> FakeStripeViewController {
        let viewController = FakeStripeViewController(viewModel: viewModel)
        return viewController
    }
    
    func updateUIViewController(_ uiViewController: FakeStripeViewController, context _: UIViewControllerRepresentableContext<FakeStripeView>) {
        if isPressed {
            uiViewController.sendBuyOrder()
        }
    }
}

class FakeStripeViewController: UIViewController {
    
    var viewModel: ViewModel?
    
    convenience init() {
        self.init(viewModel: nil)
    }
    
    init(viewModel: ViewModel?) {
        self.viewModel = viewModel
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    func sendBuyOrder() {
        // Perform Stripe calls with ViewModel data
    }
}

extension FakeStripeViewController: STPAuthenticationContext {
    func authenticationPresentingViewController() -> UIViewController {
        self
    }
}
```

**Comment by madhavSoni** [comment]
struct StripeCardTextField: UIViewRepresentable {
    @Binding var cardParams: STPPaymentMethodCardParams
    
    func makeUIView(context: Context) -> STPPaymentCardTextField {
        let cardField = STPPaymentCardTextField()
        return cardField
    }

    func updateUIView(_ uiView: STPPaymentCardTextField, context: Context) {
        DispatchQueue.main.async {
            self.cardParams = uiView.cardParams
        }
    }
}

I had something like this going and it was functional. The problem is that it caused that particular view, which is a modal view, to become incredibly buggy. Looking for alternative solutions.

**Comment by macshodan** [comment]
> > Hey everyone! It's been a while since this issue was first brought up.
> > I was able to successfully integrate stripe last year, back when I was using the early beta version of SwiftUI. Let's just say that the app didn't make the cut.
> > I started working on a new app in my free time and I am using the latest tool available, I think Apple refers to it as SwiftUI version 2 ... the only relevant change for the stripe sdk is that the app no longer uses an AppDelegate and SceneDelegate.
> > I have been fighting with it for 2 days now and I am still blocked. I raised a ticket with the support team and I am waiting to hear back from them via email.
> > Essentially, I can allow the user to select a payment method by calling the method `presentPaymentOptionsViewController()`. So far so good. I then proceed with my workflow and call `requestPayment()` and this is when the app crashes complaining that the customer context is nil.
> > After debugging for hours in a row I noticed that when the screen from `presentPaymentOptionsViewController()` is dismissed, all of the screens are reinitialised
> > I am using the same approach as @btian to get to the root view controller
> > ```
> > let keyWindow = UIApplication.shared.connectedScenes
> >             .filter({$0.activationState == .foregroundActive})
> >             .map({$0 as? UIWindowScene})
> >             .compactMap({$0})
> >             .first?.windows
> >             .filter({$0.isKeyWindow}).first
> >         self.paymentContext.hostViewController = keyWindow?.rootViewController
> > ```
> > 
> > 
> > I tried with `.filter({$0.isKeyWindow}).last` which doesn't crash the app but I am unable to type anything in the form.
> > What is interesting is that if I open the app again (re-build in XCode for example) and call `requestPayment()` **without previously calling** `presentPaymentOptionsViewController()` then I see a fulfilled payment intent in the dashboard.
> > Bottom line is that I can call `presentPaymentOptionsViewController` or `requestPayment` but not sequentially like I am supposed to (as indicated by the docs as well) because when one screen is dismissed, everything in the app reloads, causing the delegate and the customer context to be reinitialised.
> > I tried everything that I could think of to get over this and bypass the reinitialisation that occurs but I can't because there is no AppDelegate or SceneDelegate in SwiftUI 2.
> > Any help would be appreciated. Thank you!
> 
> hi amuresia
> 
> i am working to build my app on swiftui 1.0.
> i am glad you figured out to manage integrate Stripe into SwiftUi.
> 
> > Hey everyone! It's been a while since this issue was first brought up.
> > I was able to successfully integrate stripe last year, back when I was using the early beta version of SwiftUI. Let's just say that the app didn't make the cut.
> > I started working on a new app in my free time and I am using the latest tool available, I think Apple refers to it as SwiftUI version 2 ... the only relevant change for the stripe sdk is that the app no longer uses an AppDelegate and SceneDelegate.
> > I have been fighting with it for 2 days now and I am still blocked. I raised a ticket with the support team and I am waiting to hear back from them via email.
> > Essentially, I can allow the user to select a payment method by calling the method `presentPaymentOptionsViewController()`. So far so good. I then proceed with my workflow and call `requestPayment()` and this is when the app crashes complaining that the customer context is nil.
> > After debugging for hours in a row I noticed that when the screen from `presentPaymentOptionsViewController()` is dismissed, all of the screens are reinitialised
> > I am using the same approach as @btian to get to the root view controller
> > ```
> > let keyWindow = UIApplication.shared.connectedScenes
> >             .filter({$0.activationState == .foregroundActive})
> >             .map({$0 as? UIWindowScene})
> >             .compactMap({$0})
> >             .first?.windows
> >             .filter({$0.isKeyWindow}).first
> >         self.paymentContext.hostViewController = keyWindow?.rootViewController
> > ```
> > 
> > 
> > I tried with `.filter({$0.isKeyWindow}).last` which doesn't crash the app but I am unable to type anything in the form.
> > What is interesting is that if I open the app again (re-build in XCode for example) and call `requestPayment()` **without previously calling** `presentPaymentOptionsViewController()` then I see a fulfilled payment intent in the dashboard.
> > Bottom line is that I can call `presentPaymentOptionsViewController` or `requestPayment` but not sequentially like I am supposed to (as indicated by the docs as well) because when one screen is dismissed, everything in the app reloads, causing the delegate and the customer context to be reinitialised.
> > I tried everything that I could think of to get over this and bypass the reinitialisation that occurs but I can't because there is no AppDelegate or SceneDelegate in SwiftUI 2.
> > Any help would be appreciated. Thank you!
> 
> Can you put up your ready version on github helping other developer to look at it.
> 
> Bat

Hi Bat,

I assuming you're calling presentPaymentOptionsViewController() from the paymentContext object, right?
I've initialized mine in the View properties by setting it to the shared instance, so even if the view is re-initialized it will be consistent. Maybe that the issue you're having.

Here some snippets of my View:

`struct PlaceOrderView: View {
    @ObservedObject var cartModel: CartModel
    let config = STPPaymentConfiguration.shared()
    .....
`

then later in the struct I have this method implemented:

`
private func paymentContextConfiguration() {
        self.config.appleMerchantIdentifier = "your.merchant.id"
        self.config.shippingType = .delivery
        self.config.requiredBillingAddressFields = .full
        
        self.config.companyName = "yourCompanyName"
        
        self.paymentContext = STPPaymentContext(customerContext: customerContext, configuration: self.config, theme: .default())
        self.paymentContext.paymentAmount = Int(self.cartModel.cart!.totalToPay*100)
        self.paymentContext.delegate = self.cartModel
        
        
        let keyWindow = UIApplication.shared.connectedScenes
            .filter({$0.activationState == .foregroundActive})
            .map({$0 as? UIWindowScene})
            .compactMap({$0})
            .first?.windows
            .filter({$0.isKeyWindow}).first
        self.paymentContext.hostViewController = keyWindow?.rootViewController
        
        self.paymentContext.paymentCountry = "IT"
        self.paymentContext.paymentCurrency = "EUR"
    }
`

I have it tested and working under SwiftUI 2.0 with Xcode 12 against iOS 13 and iOS 14. 
The only downside is that I cannot use if from a modal view (eg. presenting from a .sheet) because it complains that it cannot present the PaymentOptionsViewController from another modal view.

**Comment by mrkvans** [comment]
Recently got Stripe working with SwiftUI for iOS 14... Here's a presentation I gave about it. [https://youtu.be/74JgD6HKg2o](url)

**Comment by davidme-stripe** [comment]
Hey all! We *are* actively watching this thread, and I understand everyone's frustration with the current state of our SwiftUI support. I have some news on the Swift front:

One of the biggest blockers of SwiftUI support is the difficulty of shipping Swift in our SDK: The Objective-C/Swift bridge has some surprising quirks, and mixed-language frameworks aren't well-supported by Xcode or our various package managers. To fix this, we're converting the entire SDK to Swift. [Here's a preview.](https://github.com/stripe/stripe-ios/issues/1680)

We're hoping this will give us a foundation on which to build Swift-first APIs and a great SwiftUI experience. Please try it out and let us know what you think!

**Comment by DrinkCupboard** [comment]
@davidme-stripe - Just wanted the chime in here and say that we're also very much looking for a SwiftUI implementation of Stripe, it'll be a vital part of our app. I believe the release of SwiftUI 2 has brought a lot of first time app builders to iOS development, and thus can only see the demand for this growing! 😊 Looking forward to this! Keep up the great work!

**Comment by amuresia** [comment]
> I assuming you're calling presentPaymentOptionsViewController() from the paymentContext object, right?
I've initialized mine in the View properties by setting it to the shared instance, so even if the view is re-initialized it will be consistent. Maybe that the issue you're having.

Thank you for chipping in @macshodan. I tried out this approach as well but I am still seeing the app crash the second time the modal to add/ edit a card is presented with 
```
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'hostViewController must not be nil on STPPaymentContext when calling pushPaymentOptionsViewController on it. Next time, set the hostViewController property first!'
```

In my view, I tried to have the following
```
let paymentConfiguration = STPPaymentConfiguration.shared()
@State var paymentContext: STPPaymentContext?
```
Then further down I called
```
    func paymentContextConfiguration() {
       ...
        let customerContext = STPCustomerContext(keyProvider: StripeApi())

        paymentContext = STPPaymentContext(customerContext: customerContext, configuration: paymentConfiguration, theme: .default())
        paymentContext!.delegate = PaymentContextDelegate()

        let keyWindow = UIApplication.shared.connectedScenes
            .filter({$0.activationState == .foregroundActive})
            .map({$0 as? UIWindowScene})
            .compactMap({$0})
            .first?.windows
            .filter({$0.isKeyWindow}).first

        paymentContext!.hostViewController = keyWindow?.rootViewController
    }
```

Please let me know if I misunderstood your suggestion, I'm really keen on getting this to work properly.

**Comment by macshodan** [comment]
Hi @amuresia, I believe the problem is that you're initializing the paymentContext @state object with nullable. 
In my code I did the following for the view properties:

```
let config = STPPaymentConfiguration.shared()
@State private var paymentContext: STPPaymentContext!
let customerContext = STPCustomerContext(keyProvider: NetworkAPI.shared)
```

where the NetworkAPI class implements the STPCustomerEphemeralKeyProvider delegate.
Then in the paymentContextConfiguration I initialize my paymentContext with view customerContex. 
Apart from that I don't see any other differences in our code.

Hope this helps.

**Comment by madhavSoni** [comment]
For some reason, I'm having an issue making direct charges in Stripe connect with a Swift UI integration.  I formerly used transfers and they worked successfully, but switching to direct charges seemed to cause failures in my integration. 

I'm getting the following error whenever I try doing a direct charge.

_The PaymentIntent requires a payment method._

Anyone run into this issue or know what's going on? The only difference in my code is in my backend. 

I formerly used this code because it worked:
```
  intent = stripe.PaymentIntent.create(
      amount=amount,
      currency='usd',
      on_behalf_of=account_id,
      receipt_email=email,
      transfer_data={
          'destination': account_id,
      },
      application_fee_amount=app_fee
  )
```

However, now my integration requires me to use direct charges, so now it looks something more like this:
```
   intent = stripe.PaymentIntent.create(
        amount=amount,
        currency='usd',
         payment_method_types=['card'],
        receipt_email=email,
        stripe_account= account_id,
        application_fee_amount=app_fee
    )
```

Anyone have any idea why this change is no longer attaching the payment method? I saw in the Stripe docs that it is necessary to clone a payment method for connect: 

https://stripe.com/docs/payments/payment-methods/connect#creating-paymentmethods-directly-on-the-connected-account

However, there isn't an explanation for how to do this in iOS, yet alone SwiftUI.

Any help would be appreciated as this is the last part of my integration!
Thanks

**Comment by aliriaz-stripe** [comment]
@madhavSoni hello!

We should take this to a different channel and keep this thread for SwiftUI specific questions/comments, I'd recommend writing in to Support at https://support.stripe.com/contact with your code snippets and we can help you out there!

Could you include in your message, how you're initializing `stripe-ios` in your mobile app? For Direct Charges, your client-app needs to be initialized with both your Platform publishable key, _and_ the Stripe account ID of the Connect account you are creating the Direct Charge on.

Please include that code snippet showing how you're initializing Stripe in your iOS app, the above code snippet and any request IDs for the PaymentIntent creation and confirmation when you write in to Support and we can help investigate!

**Comment by dqii** [comment]
It looks like converting the SDK to Swift was completed. Has there been progress on support SwiftUI 2?

**Comment by ShahinBorujeni** [comment]
Any update on support for SwiftUI? Our app uses SwiftUI  and we need to implement Stripe to finish it up! if anyone has any solutions I'd appreciate it if you share your workaround! this is vital for our business. Thanks

**Comment by mrkvans** [comment]
Hi @ShahinBorujeni I recently launched a 100% SwiftUI app that leveraged Stripe. 
I didn't use any Stripe UI, I used Apple Pay UI with the Stripe payment processing. 

What type of payment UI are you trying to implement?

**Comment by ShahinBorujeni** [comment]
We wanted to use the prebuilt UI for adding credit cards and making payment. But since there’s no official support for prebuilt on SwiftUI I will watch your YouTube video to at least implement Apple Pay for now. Thank you for the response

**Comment by fassko** [comment]
@ShahinBorujeni I'm using Stripe UI with SwiftUI. I'm stitching together and what I do is to replace the main view controller with the Stripe UI one. Check this one https://github.com/nelglez/stripe-swiftui I used this as an example, maybe it helps.

**Comment by davidme-stripe** [comment]
Hi all! We plan to launch SwiftUI support as part of our new [Mobile Payments UI](https://stripe.com/docs/mobile/payments-ui-beta-1-mc2ps), which is currently in beta.

For folks on this thread, I've [added a preview to the `mc-beta-1-swiftui` branch](https://github.com/stripe/stripe-ios/tree/mc-beta-1-swiftui). We haven't updated the beta docs for SwiftUI yet, but we have a few SwiftUI examples to serve as a reference:
* [`ExampleSwiftUIPaymentSheet.swift`](https://github.com/stripe/stripe-ios/blob/mc-beta-1-swiftui/Example/PaymentSheet%20Example/PaymentSheet%20Example/ExampleSwiftUIPaymentSheet.swift) demonstrates our standard [PaymentSheet flow](https://stripe.com/docs/mobile/payments-ui-beta-1-mc2ps#app).
* If you'd prefer more customization, [`ExampleSwiftUICustomPaymentFlow.swift`](https://github.com/stripe/stripe-ios/blob/mc-beta-1-swiftui/Example/PaymentSheet%20Example/PaymentSheet%20Example/ExampleSwiftUICustomPaymentFlow.swift) demonstrates a [custom flow](https://stripe.com/docs/mobile/payments-ui-beta-1-mc2ps#custom).

You can run the `PaymentSheet Example` app from [`Stripe.xcworkspace`](https://github.com/stripe/stripe-ios/tree/mc-beta-1-swiftui) to try each flow.

Please also [register for our beta program](https://forms.gle/76kkxzqXxdyptPf67) — we'll use that mailing list to collect feedback and notify you of new beta releases. You can also share feedback with us by emailing mobile-payments-ui-beta@stripe.com.

**Comment by DrinkCupboard** [comment]
@davidme-stripe This is brilliant news! Thanks for your dedication on this! 🎉👏 P.S. We just signed up!

**Comment by davidme-stripe** [comment]
Update: We've released [beta 2](https://github.com/stripe/stripe-ios/tree/mc-beta-2) and added some initial [SwiftUI docs](https://stripe.com/docs/mobile/payments-ui-beta).

Again, please [register for our beta program](https://forms.gle/76kkxzqXxdyptPf67) if you haven't!

**Comment by davidme-stripe** [comment]
In [21.3.0](https://github.com/stripe/stripe-ios/releases/tag/21.3.0), we added a SwiftUI [`STPPaymentCardTextField.Representable`](https://stripe.dev/stripe-ios/docs/Classes/STPPaymentCardTextField/Representable.html) and [ViewModifiers for confirming a PaymentIntent or SetupIntent](https://stripe.dev/stripe-ios/docs/Extensions/View.html). An example is available in the [IntegrationTester](https://github.com/stripe/stripe-ios/blob/21.3.0/IntegrationTester/IntegrationTester/Examples/CustomCard.swift) app.

We still recommend using the [Mobile Payments UI](https://stripe.com/docs/mobile/payments-ui-beta) for new apps, but these new wrappers could be useful if you're trying to convert an existing integration to native SwiftUI.

**Comment by ilazakis** [comment]
Is there an `STPAddCardViewController` equivalent in the Mobile Payments UI beta. I’m new to the SDK, but it looks like it’s the only way if one only wants to add a Payment Method for later use?

**Comment by davidme-stripe** [comment]
@ilazakis, it sounds like you're looking for [SetupIntent](https://stripe.com/docs/payments/setup-intents) support. This isn't included yet, but we intend to add it before the new UI is out of beta.

**Comment by ilazakis** [comment]
Thank you @davidme-stripe! 

Would working with `STPPaymentCardTextField.Representable` directly still keep us safely within PCI compliance or is one required to use `STPAddCardViewController`? Happy to keep using a navigation bar hack others have used recently with `STPAddCardViewController` to create payment methods until the newer version lands. 

Phrased differently, is using `STPPaymentCardTextField.Representable` and `apiClient.createPaymentMethod` ok as far as PCI compliance is concerned? Or is use of `STPAddCardViewController` mandatory? Trying to avoid `SetupIntent` if possible.

**Comment by davidme-stripe** [comment]
The Mobile SDK section of our [security guide](https://stripe.com/docs/security/guide) is the best source for information on this. I would treat the `STPPaymentCardTextField -> STPAPIClient.shared.createPaymentMethod()` flow as equivalent to `STPAddCardViewController`, as long as you pass an `STPPaymentMethodCardParams` and avoid handling or introspecting the raw card information.

**Comment by ilazakis** [comment]
Thanks David! 

That's exactly the setup I'm currently using as an alternative, following `STPAddCardViewController`'s implementation really closely, effectively copying it as-is.

**Comment by davidme-stripe** [comment]
We've released [21.5.0](https://github.com/stripe/stripe-ios/releases/tag/21.5.0) with a public beta of PaymentSheet, including support for PaymentIntents and SetupIntents. Between [PaymentSheet](https://stripe.com/docs/mobile/payments-ui-beta), [STPPaymentCardTextField](https://stripe.dev/stripe-ios/docs/Classes/STPPaymentCardTextField/Representable.html), and [.paymentConfirmationSheet](https://stripe.dev/stripe-ios/docs/Extensions/View.html#/s:7SwiftUI4ViewP6StripeE24paymentConfirmationSheet19isConfirmingPayment0E12IntentParams12onCompletionQrAA7BindingVySbG_AD010STPPaymentkL0CyAD0P19HandlerActionStatusO_AD0pK0CSgSo7NSErrorCSgtctF), I think we've covered everyone's SwiftUI needs. We intend to provide native SwiftUI support or SwiftUI wrappers for all future UI surfaces.

Thanks to everyone in this thread for your feedback! If you have any other questions or suggestions, feel free to file a new issue.

**Comment by ShahinBorujeni** [comment]
Hey David,
It's so refreshing to see stripe on SwifUI, Thanks for all the hard work. I noticed a bug while working with Prebuilt UI Inside of a modal(.sheet or .fullScreenCover) that prevents the bottom payment sheet to appear. Can you let me know if there is a workaround for this?

**Comment by codeonomics** [comment]
Would be good to see SwiftUI documentation for connecting a Stripe Account via SafariDelegate in the docs: https://stripe.com/docs/connect/collect-then-transfer-guide?platform=ios&ui=payment-sheet#redirect-your-user-to-the-account-link-url - only UIKit currently.
